File: colonization-chaos/solutions/sol.py
n, m = map(int, input().split())
resources = [int(x) for x in input().split()]
planets = [int(x) for x in input().split()]

dp = {0}
highest_in_planets = max(planets)
for value in resources:
    for existing in list(dp):
        if existing + value <= highest_in_planets:
            dp.add(existing + value)
print(sum(total in dp for total in planets))

File: colonization-chaos/solutions/sol.java
import java.util.*;

public class Solution {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        int[] resources = new int[n];
        for (int i = 0; i < n; i++) {
            resources[i] = sc.nextInt();
        }

        int[] planets = new int[m];
        for (int i = 0; i < m; i++) {
            planets[i] = sc.nextInt();
        }

        Set<Integer> dp = new HashSet<>();
        dp.add(0);
        int highest_in_planets = Arrays.stream(planets).max().getAsInt();
        for (int value : resources) {
            Set<Integer> dp_copy = new HashSet<>(dp);
            for (int existing : dp_copy) {
                if (existing + value <= highest_in_planets) {
                    dp.add(existing + value);
                }
            }
        }
        int count = 0;
        for (int total : planets) {
            if (dp.contains(total)) {
                count++;
            }
        }
        System.out.println(count);
    }
}